File: crates/zer0-core/src/ext/compat.js
// zer0 compatibility file.
//
// This file is not part of the extension that ships alongside it. zer0 writes
// it during install and points `background` at it, so it runs before the
// extension's own code; `manifest.json` records both facts under `zer0_compat`,
// and the extension's own entry point is re-entered at the bottom of this file.
//
// It exists because a member missing from a namespace that exists is fatal
// under MV3, and there is nothing an extension can do about it. Measured across
// 59 store packages: 14 background workers never start, and six of those die on
// a single missing member — React Developer Tools on
// `chrome.scripting.ExecutionWorld.ISOLATED`, two strings.
//
// ## What is allowed in here
//
// 1. **Enums and constants whose value Chrome documents as a literal.** There
//    is nothing to implement and nothing to get wrong, because the enum *is*
//    its value.
// 2. **Event objects that only need to exist.** MV3 requires listeners to be
//    registered synchronously while the worker starts, so a worker that listens
//    for something this engine lacks is dead before it does anything. An
//    `addListener` that registers and never fires never claims the event
//    happened.
// 3. **`chrome.storage.managed`.** zer0 has no enterprise policy mechanism, so
//    managed storage is empty — not "not implemented", empty. Reading it
//    answers that; writing to it rejects, which is what Chrome does too.
//
// 4. **Methods zer0 really carries out**, over its own channel — the fourth
//    tier, added by ADR-0103, and the one to be most careful about. Nothing is
//    here unless this browser really does the thing: `chrome.downloads` is the
//    download subsystem ADR-0027 and ADR-0101 already built, and
//    `chrome.idle.queryState` is a number the system hands over. Every call
//    goes to `crates/zer0-core/src/extension_api.rs`, which decides — including
//    deciding to refuse.
// 5. **`chrome.management.getSelf`**, computed from the extension's own
//    manifest. It needs no permission in Chrome either and asserts nothing this
//    file does not already have in its hands.
//
// ## What is not, and why
//
// **Anything that would have to do something zer0 does not do.**
// `tabs.executeScript`, `runtime.getContexts`, `i18n.detectLanguage`,
// delivering a notification. A method that resolves without doing the thing is
// a silent failure nobody can diagnose, and it is strictly worse than the loud
// one it replaced (ADR-0077). The fourth tier above is the *opposite* of that
// and stays the opposite only while every method in it is backed: adding one
// that answers out of this file rather than out of the browser is how this
// becomes the thing it was written to prevent.
//
// **Capacity numbers.** `MAX_NUMBER_OF_DYNAMIC_RULES` and its neighbours are
// claims about what *this* engine will accept. Chrome's number is not evidence
// about WebKit's, and stating one would be the browser saying something it
// cannot prove (ADR-0018). Identifiers like `DYNAMIC_RULESET_ID` are not
// capacities and are stated.
//
// **Type names.** `chrome.tabs.Tab` and friends are documentation, undefined in
// Chrome too. A survey that reads a package's source finds them; defining them
// would be inventing surface Chrome does not have.
//
// ## The two rules that keep it from becoming a lie
//
// **Nothing is installed over something that exists.** Every write below is
// guarded, so the day WebKit implements one of these the real one wins and this
// file goes quiet on its own, without anybody re-measuring.
//
// **Nothing is invented.** A member this file does not list stays `undefined`.
// That is deliberate and it is the opposite of the obvious design: a stub
// function that throws when called reads as more helpful, and it would defeat
// `if (chrome.notifications.create)` — the extension's own way of finding out
// and taking another path — while asserting a method exists that does not.
// Leaving it undefined keeps that check honest, and a call to it still fails
// loudly, at the call site, instead of at startup.

(function () {
  'use strict';

  // A registration that is remembered and never fired. `hasListener` answers
  // about registration, which is a fact, and not about delivery, which is not.
  function eventObject() {
    var listeners = [];
    return {
      addListener: function (fn) {
        if (typeof fn === 'function' && listeners.indexOf(fn) === -1) listeners.push(fn);
      },
      removeListener: function (fn) {
        var at = listeners.indexOf(fn);
        if (at !== -1) listeners.splice(at, 1);
      },
      hasListener: function (fn) {
        return listeners.indexOf(fn) !== -1;
      },
      hasListeners: function () {
        return listeners.length > 0;
      },
    };
  }

  // Chrome's documented literals, keyed by the path they live at.
  var ENUMS = {
    'runtime.ContextType': {
      TAB: 'TAB',
      POPUP: 'POPUP',
      BACKGROUND: 'BACKGROUND',
      OFFSCREEN_DOCUMENT: 'OFFSCREEN_DOCUMENT',
      SIDE_PANEL: 'SIDE_PANEL',
      DEVELOPER_TOOLS: 'DEVELOPER_TOOLS',
    },
    'runtime.OnInstalledReason': {
      INSTALL: 'install',
      UPDATE: 'update',
      CHROME_UPDATE: 'chrome_update',
      SHARED_MODULE_UPDATE: 'shared_module_update',
    },
    'runtime.OnRestartRequiredReason': {
      APP_UPDATE: 'app_update',
      OS_UPDATE: 'os_update',
      PERIODIC: 'periodic',
    },
    'runtime.PlatformArch': {
      ARM: 'arm',
      ARM64: 'arm64',
      X86_32: 'x86-32',
      X86_64: 'x86-64',
      MIPS: 'mips',
      MIPS64: 'mips64',
    },
    'runtime.PlatformOs': {
      MAC: 'mac',
      WIN: 'win',
      ANDROID: 'android',
      CROS: 'cros',
      LINUX: 'linux',
      OPENBSD: 'openbsd',
      FUCHSIA: 'fuchsia',
    },
    'scripting.ExecutionWorld': {
      ISOLATED: 'ISOLATED',
      MAIN: 'MAIN',
    },
    'declarativeNetRequest.DomainType': {
      FIRST_PARTY: 'firstParty',
      THIRD_PARTY: 'thirdParty',
    },
    'declarativeNetRequest.HeaderOperation': {
      APPEND: 'append',
      SET: 'set',
      REMOVE: 'remove',
    },
    'declarativeNetRequest.ResourceType': {
      MAIN_FRAME: 'main_frame',
      SUB_FRAME: 'sub_frame',
      STYLESHEET: 'stylesheet',
      SCRIPT: 'script',
      IMAGE: 'image',
      FONT: 'font',
      OBJECT: 'object',
      XMLHTTPREQUEST: 'xmlhttprequest',
      PING: 'ping',
      CSP_REPORT: 'csp_report',
      MEDIA: 'media',
      WEBSOCKET: 'websocket',
      WEBTRANSPORT: 'webtransport',
      WEBBUNDLE: 'webbundle',
      OTHER: 'other',
    },
    'declarativeNetRequest.RuleActionType': {
      BLOCK: 'block',
      REDIRECT: 'redirect',
      ALLOW: 'allow',
      UPGRADE_SCHEME: 'upgradeScheme',
      MODIFY_HEADERS: 'modifyHeaders',
      ALLOW_ALL_REQUESTS: 'allowAllRequests',
    },
    'declarativeNetRequest.UnsupportedRegexReason': {
      SYNTAX_ERROR: 'syntaxError',
      MEMORY_LIMIT_EXCEEDED: 'memoryLimitExceeded',
    },
    // The whole `On*Options` family, because they are one shape and an
    // extension reaching for the one we happened to measure would have reached
    // for its neighbour just as easily. DuckDuckGo dies on
    // `OnHeadersReceivedOptions.EXTRA_HEADERS`.
    'webRequest.OnBeforeRequestOptions': {
      BLOCKING: 'blocking',
      REQUEST_BODY: 'requestBody',
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.OnBeforeSendHeadersOptions': {
      REQUEST_HEADERS: 'requestHeaders',
      BLOCKING: 'blocking',
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.OnSendHeadersOptions': {
      REQUEST_HEADERS: 'requestHeaders',
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.OnHeadersReceivedOptions': {
      BLOCKING: 'blocking',
      RESPONSE_HEADERS: 'responseHeaders',
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.OnAuthRequiredOptions': {
      RESPONSE_HEADERS: 'responseHeaders',
      BLOCKING: 'blocking',
      ASYNC_BLOCKING: 'asyncBlocking',
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.OnResponseStartedOptions': {
      RESPONSE_HEADERS: 'responseHeaders',
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.OnBeforeRedirectOptions': {
      RESPONSE_HEADERS: 'responseHeaders',
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.OnCompletedOptions': {
      RESPONSE_HEADERS: 'responseHeaders',
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.OnErrorOccurredOptions': {
      EXTRA_HEADERS: 'extraHeaders',
    },
    'webRequest.ResourceType': {
      MAIN_FRAME: 'main_frame',
      SUB_FRAME: 'sub_frame',
      STYLESHEET: 'stylesheet',
      SCRIPT: 'script',
      IMAGE: 'image',
      FONT: 'font',
      OBJECT: 'object',
      XMLHTTPREQUEST: 'xmlhttprequest',
      PING: 'ping',
      CSP_REPORT: 'csp_report',
      MEDIA: 'media',
      WEBSOCKET: 'websocket',
      WEBTRANSPORT: 'webtransport',
      WEBBUNDLE: 'webbundle',
      OTHER: 'other',
    },
  };

  // Identifiers, not capacities. Chrome documents these two as the literal
  // names of the two rulesets every extension has.
  var CONSTANTS = {
    'declarativeNetRequest.DYNAMIC_RULESET_ID': '_dynamic',
    'declarativeNetRequest.SESSION_RULESET_ID': '_session',
  };

  // Event objects. Every one of these was measured absent from a namespace that
  // is present, in a package that registers it while its worker starts.
  var EVENTS = [
    'action.onUserSettingsChanged',
    'declarativeNetRequest.onRuleMatchedDebug',
    'extension.onRequest',
    'extension.onRequestExternal',
    'runtime.onBrowserUpdateAvailable',
    'runtime.onPerformanceWarning',
    'runtime.onRestartRequired',
    'runtime.onSuspend',
    'runtime.onSuspendCanceled',
    'runtime.onUpdateAvailable',
    'runtime.onUserScriptConnect',
    'runtime.onUserScriptMessage',
    'webNavigation.onCreatedNavigationTarget',
    'webNavigation.onHistoryStateUpdated',
    'webNavigation.onReferenceFragmentUpdated',
    'windows.onBoundsChanged',
    // The three that need the namespace below to exist at all.
    'notifications.onButtonClicked',
    'notifications.onClicked',
    'notifications.onClosed',
  ];

  // The namespaces this file creates rather than adds to.
  //
  // `notifications` is the original and the whole reason this list exists:
  // 1Password, Redux DevTools and Violentmonkey each call
  // `notifications.onClicked.addListener` at the top level of their worker,
  // which MV3 makes fatal. What gets created carries the three event objects
  // above and nothing else — `create` stays undefined, so an extension that
  // checks for it takes its own other path, and putting a notification on
  // somebody's screen is a permission decision zer0 has not taken (ADR-0103).
  //
  // `downloads`, `idle` and `management` are here because zer0 answers them
  // itself. WebKit installs none of the three, so on this engine every write
  // below is the first one; the guard in `define` still stands, so the day it
  // does install one, WebKit's wins.
  var CREATABLE = ['notifications', 'downloads', 'idle', 'management'];

  // MARK: - The channel

  // How this file reaches the browser. Measured on macOS 26.6: a `fetch` from a
  // background service worker to a scheme the extension controller's
  // configuration has a handler for arrives at that handler, with the body and
  // with an `Origin` naming which extension asked. It is the only road out of a
  // worker that does not need the `nativeMessaging` permission, which most of
  // these packages never asked for.
  var CALL_URL = 'zer0-extension-api://call/';

  // A refusal from the browser reaches the caller as a rejection, and a
  // callback caller gets it thrown.
  //
  // `chrome.runtime.lastError` is what Chrome would set, and this file has no
  // way to set it — the same wall `storage.managed` hit above, answered the
  // same way. Throwing is the loud answer available here, and it is loud at the
  // call site with the browser's own sentence in it, rather than the "is not a
  // function" a missing member would give.
  function call(method, args) {
    return fetch(CALL_URL + method, {
      method: 'POST',
      body: JSON.stringify(args === undefined ? null : args),
    })
      .then(function (response) {
        return response.json();
      })
      .then(function (answered) {
        if (answered && typeof answered.error === 'string') {
          throw new Error(answered.error);
        }
        if (!answered || !('ok' in answered)) {
          throw new Error('zer0 answered ' + method + ' with something unreadable.');
        }
        return answered.ok;
      });
  }

  // Chrome's two calling conventions over one channel. A trailing function is a
  // callback, anything else is a promise.
  //
  // The callback branch does **not** swallow the failure: there is no
  // `lastError` to put it in, so it becomes an unhandled rejection, which is
  // loud in the console and does not take the worker down. Calling the callback
  // with `undefined` would be this file reporting success.
  function bridged(method) {
    return function () {
      var args = Array.prototype.slice.call(arguments);
      var callback = typeof args[args.length - 1] === 'function' ? args.pop() : null;
      var answered = call(method, args[0]);
      if (!callback) return answered;
      answered.then(function (value) {
        callback(value);
      });
      return undefined;
    };
  }

  // Every call, including the two the browser only ever refuses.
  //
  // `pause` and `resume` go down the same road as the rest and come back with
  // the reason zer0 will not do them (`extension_api.rs`). Defining them at all
  // is a departure from "nothing is invented" worth being explicit about: this
  // namespace is zer0's, so `if (chrome.downloads.pause)` no longer tells an
  // extension anything about the engine — it would only tell it about this
  // table. A refusal carrying the browser's reason is more use than a silence
  // carrying none.
  //
  // Nothing else Chrome documents on these namespaces is here.
  // `downloads.setShelfEnabled`, `downloads.acceptDanger`,
  // `idle.setDetectionInterval` and the rest stay undefined, so an extension
  // checking for one still finds out honestly, and calling one still fails at
  // the call site.
  var CALLS = {
    'downloads.download': bridged('downloads.download'),
    'downloads.search': bridged('downloads.search'),
    'downloads.cancel': bridged('downloads.cancel'),
    'downloads.erase': bridged('downloads.erase'),
    'downloads.open': bridged('downloads.open'),
    'downloads.show': bridged('downloads.show'),
    'downloads.pause': bridged('downloads.pause'),
    'downloads.resume': bridged('downloads.resume'),
    'idle.queryState': bridged('idle.queryState'),
  };

  // Events on namespaces zer0 answers, which exist and never fire.
  //
  // The same tier-2 bargain as everywhere else in this file, and the same cost:
  // an extension whose only road is `downloads.onChanged` now starts and then
  // does nothing, which is better than dead and is not working. There is no
  // road from this browser to a running worker to fire them down — a worker
  // that is asleep is asleep, and holding a `fetch` open to keep one awake
  // would be this file deciding an extension may never be suspended.
  var ANSWERED_EVENTS = [
    'downloads.onCreated',
    'downloads.onChanged',
    'downloads.onErased',
    'downloads.onDeterminingFilename',
    'idle.onStateChanged',
  ];

  // `chrome.management.getSelf`, out of the extension's own manifest.
  //
  // Chrome needs no permission for this one and neither does zer0, because
  // nothing here is a fact the extension did not already hold: it is its own
  // `manifest.json` and its own id, rearranged. Everything else in
  // `chrome.management` — `getAll`, `setEnabled`, `uninstall` — stays undefined
  // and is refused on the Extensions screen in as many words: one extension does
  // not get to switch off another.
  function selfDescription() {
    var manifest = {};
    try {
      manifest = chrome.runtime.getManifest() || {};
    } catch (e) {
      return null;
    }
    return {
      id: chrome.runtime.id,
      name: manifest.name || '',
      shortName: manifest.short_name || manifest.name || '',
      description: manifest.description || '',
      version: manifest.version || '',
      versionName: manifest.version_name,
      mayDisable: true,
      enabled: true,
      installType: 'normal',
      type: 'extension',
      permissions: (manifest.permissions || []).slice(),
      hostPermissions: (manifest.host_permissions || []).slice(),
      icons: [],
      offlineEnabled: !!manifest.offline_enabled,
      optionsUrl: '',
      homepageUrl: manifest.homepage_url,
      updateUrl: manifest.update_url,
      isApp: false,
    };
  }

  function namespaceIn(root, name) {
    var existing = root[name];
    if (existing !== undefined && existing !== null) return existing;
    if (CREATABLE.indexOf(name) === -1) return null;
    try {
      root[name] = {};
    } catch (e) {
      return null;
    }
    return root[name] || null;
  }

  function define(root, path, value) {
    var dot = path.indexOf('.');
    var namespace = namespaceIn(root, path.slice(0, dot));
    if (!namespace) return;
    var member = path.slice(dot + 1);
    if (namespace[member] !== undefined) return;
    try {
      namespace[member] = value;
    } catch (e) {
      // A frozen namespace is WebKit's to decide. Nothing else to try, and
      // nothing worth reporting: the extension is exactly as it was.
    }
  }

  // `chrome.storage.managed`, empty and read-only.
  //
  // Twelve of 59 packages read it. Empty is the correct answer rather than a
  // stand-in for one: managed storage holds what an enterprise policy put
  // there, zer0 has no mechanism by which a policy could put anything there,
  // so there is nothing in it and there is no version of this browser where
  // there would be. Chrome makes it read-only too, so the rejection below is
  // Chrome's behaviour and not zer0 refusing something.
  function managedStorage() {
    // Chrome's `get` answers with the defaults it was handed, since nothing is
    // ever set. A string or an array of keys asks for values that do not exist
    // and gets an empty object.
    function readBack(keys) {
      var out = {};
      if (keys && typeof keys === 'object' && !Array.isArray(keys)) {
        for (var key in keys) {
          if (Object.prototype.hasOwnProperty.call(keys, key)) out[key] = keys[key];
        }
      }
      return out;
    }

    function answer(args, value) {
      var last = args.length ? args[args.length - 1] : undefined;
      if (typeof last === 'function') {
        last(value);
        return undefined;
      }
      return Promise.resolve(value);
    }

    function readOnly(method) {
      return function () {
        var error = new Error(
          'chrome.storage.managed is read-only, so ' + method + ' cannot change it.'
        );
        var last = arguments.length ? arguments[arguments.length - 1] : undefined;
        if (typeof last === 'function') {
          // A callback caller reads `runtime.lastError`, which this file has no
          // way to set. Throwing is the loud answer available here, and it is
          // loud at the call site rather than at startup.
          throw error;
        }
        return Promise.reject(error);
      };
    }

    return {
      get: function (keys) {
        return answer(arguments, readBack(keys));
      },
      getBytesInUse: function () {
        return answer(arguments, 0);
      },
      getKeys: function () {
        return answer(arguments, []);
      },
      set: readOnly('set'),
      remove: readOnly('remove'),
      clear: readOnly('clear'),
      onChanged: eventObject(),
    };
  }

  function apply(root) {
    if (!root) return;

    var path;
    for (path in ENUMS) {
      if (Object.prototype.hasOwnProperty.call(ENUMS, path)) define(root, path, ENUMS[path]);
    }
    for (path in CONSTANTS) {
      if (Object.prototype.hasOwnProperty.call(CONSTANTS, path)) {
        define(root, path, CONSTANTS[path]);
      }
    }
    for (var i = 0; i < EVENTS.length; i++) {
      define(root, EVENTS[i], eventObject());
    }

    // What zer0 answers itself. Guarded like everything else, so an engine that
    // ships one of these keeps its own.
    for (path in CALLS) {
      if (Object.prototype.hasOwnProperty.call(CALLS, path)) define(root, path, CALLS[path]);
    }
    for (var j = 0; j < ANSWERED_EVENTS.length; j++) {
      define(root, ANSWERED_EVENTS[j], eventObject());
    }

    // Only where the extension can be asked what it is. In a context with no
    // `chrome.runtime` there is nothing to answer out of, and a `getSelf` that
    // answered anyway would be inventing the description.
    var described = selfDescription();
    if (described) {
      define(root, 'management.getSelf', function (callback) {
        if (typeof callback === 'function') {
          callback(described);
          return undefined;
        }
        return Promise.resolve(described);
      });
    }

    // Only where `chrome.storage` is there at all. A permission the person
    // withheld leaves the whole namespace undefined (ADR-0077), and handing an
    // extension a storage it was refused would undo that decision.
    define(root, 'storage.managed', managedStorage());
  }

  var chromeRoot = typeof chrome !== 'undefined' ? chrome : null;
  var browserRoot = typeof browser !== 'undefined' ? browser : null;
  apply(chromeRoot);
  if (browserRoot && browserRoot !== chromeRoot) apply(browserRoot);
})();

File: crates/zer0-core/src/ext/compat.rs
//! The one modification zer0 makes to somebody else's package.
//!
//! ADR-0020 says there is no shim layer and no `chrome.*` polyfill of our own,
//! and that was the right call for a *namespace*: what WebKit implements is
//! Apple's to widen and reimplementing it is a treadmill. It is the wrong call
//! for a **member missing from a namespace that exists**, which is a different
//! failure with a different cost — measured across 59 store packages, six
//! background workers die on one absent member each, and React Developer Tools
//! dies on `chrome.scripting.ExecutionWorld.ISOLATED`, which is two strings.
//!
//! ## Why this has to happen on disk
//!
//! Measured, on macOS 26.6, for reaching an extension's background worker:
//!
//! | Route | Reaches the worker |
//! | --- | --- |
//! | `WKWebExtensionController.Configuration.webViewConfiguration.userContentController` + `WKUserScript` | no |
//! | `WKWebExtensionContext.unsupportedAPIs` | no — it only removes |
//! | rewriting `background` in `manifest.json` at unpack | yes |
//!
//! There is no third option. So the unpacked tree stops being a copy of the
//! archive, and ADR-0022 and ADR-0024 both care about that. What answers them
//! is that the modification is **bounded, named and recorded in the file it
//! modifies**: one file is added — two for a module worker, see
//! [`COMPAT_API_FILE`] — under names no extension would choose, and
//! `manifest.json` carries a `zer0_compat` block saying so and naming where the
//! extension's own code still begins. A person can read both without a hex
//! editor, and the Extensions screen prints them.
//!
//! ## What is not kept
//!
//! Not the verified CRX. Holding a second copy of every package — 361 MB for
//! AdBlock alone — buys a comparison nothing performs and nobody is offered.
//! The auditable thing is what the modification *is*, and that is two files
//! with fixed names plus two keys in a manifest that names the entry point it
//! replaced. That is readable; a blob nobody diffs is not.

use std::fs;
use std::path::Path;

use super::ExtError;

/// The file zer0 adds and points `background` at. Named so it cannot be
/// mistaken for the extension's own.
pub const COMPAT_FILE: &str = "zer0-compat.js";

/// The second file, added only for a **module** service worker.
///
/// One file would do if a module could run something before its own imports.
/// It cannot: static `import` is hoisted and evaluated before the importing
/// module's body, so a shim written after one runs second — too late for a
/// worker that has already died. And the obvious escape, `await import()`, is
/// not available at all: measured on macOS 26.6, a module service worker
/// reaches it and throws *"Dynamic-import is not available in Worklets or
/// ServiceWorkers"*. That version of this file shipped and looked like it
/// worked, because the worker came up clean and the extension's own code never
/// ran — the silent failure this whole change is forbidden to create.
///
/// So a module worker gets an entry that is nothing but two static imports, in
/// order: this file, then the extension's. Which is also strictly better than
/// the top-level `await` it replaced, since nothing is deferred past the
/// startup turn MV3 requires listeners to be registered in.
pub const COMPAT_API_FILE: &str = "zer0-compat-api.js";

/// The manifest key recording that this package was modified and by what.
pub const MANIFEST_KEY: &str = "zer0_compat";

/// What gets written into every package that has a background this can reach.
///
/// `include_str!` rather than a string here: the file is JavaScript, it is read
/// and argued about as JavaScript, and the Swift suite evaluates these exact
/// bytes in a real interpreter to hold it to what it claims.
pub const SOURCE: &str = include_str!("compat.js");

/// That a package was modified, and what by.
///
/// One record carrying both halves, so a screen cannot draw the fact that the
/// package was changed without also being handed the names of the files that
/// changed it and the entry point they run in front of.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "ffi", derive(uniffi::Record))]
pub struct CompatNotice {
    /// Every file zer0 added, relative to the package root. One for a classic
    /// worker or an MV2 background, two for a module worker — see
    /// [`COMPAT_API_FILE`] for why the second one has to exist.
    pub added_files: Vec<String>,
    /// Where the extension's own background code begins, and still begins.
    pub original_entry_point: String,
}

/// The background shape a package declares, as far as this cares.
enum Background {
    /// MV3. `module` decides how the original is re-entered, and the two ways
    /// are not interchangeable in either direction.
    Worker { path: String, module: bool },
    /// MV2. Ours goes in front of the list rather than replacing it.
    Scripts { first: String },
}

/// Add the compatibility file to an unpacked package.
///
/// Runs inside the staging directory, before the swap, so a failure here leaves
/// nothing behind and leaves any working version exactly where it was
/// (ADR-0022).
///
/// **Refuses rather than repairs.** A background this cannot reach — an HTML
/// background page, no background at all, an entry point that is not a plain
/// path inside the package — is left exactly as it arrived, with no file added
/// and no key written. There is no half-modified state to explain.
pub fn inject(dir: &Path) -> Result<(), ExtError> {
    let manifest_path = dir.join("manifest.json");
    if !manifest_path.exists() {
        // `read_manifest` is the one that refuses a package with no manifest,
        // and it says so in the error a person reads. Nothing to do here.
        return Ok(());
    }

    // A package that already ships a file by either of these names is one we do
    // not understand, and overwriting somebody's file to install a
    // compatibility layer is the kind of repair that becomes a bug report about
    // them.
    if dir.join(COMPAT_FILE).exists() || dir.join(COMPAT_API_FILE).exists() {
        return Ok(());
    }

    let json = super::read_package_json(&manifest_path)?;
    let Ok(serde_json::Value::Object(mut manifest)) = serde_json::from_str(&json) else {
        // Unparseable, or not an object. `manifest::parse` is where that
        // becomes an error worth reading.
        return Ok(());
    };

    let Some(background) = read_background(manifest.get("background")) else {
        return Ok(());
    };

    let entry = match &background {
        Background::Worker { path, .. } => path.clone(),
        Background::Scripts { first } => first.clone(),
    };

    let mut added = vec![COMPAT_FILE.to_string()];
    match &background {
        Background::Worker {
            path, module: true, ..
        } => {
            // Two static imports, in the order they are written. The extension's
            // own module is the second, so everything below has already run by
            // the time its first line does.
            fs::write(
                dir.join(COMPAT_FILE),
                format!(
                    "// Added by zer0. {COMPAT_API_FILE} is the compatibility file; the\n\
                     // second import is where this extension's own code begins.\n\
                     import {};\nimport {};\n",
                    module_specifier(COMPAT_API_FILE),
                    module_specifier(path)
                ),
            )?;
            fs::write(dir.join(COMPAT_API_FILE), SOURCE)?;
            added.push(COMPAT_API_FILE.to_string());
        }
        Background::Worker {
            path,
            module: false,
            ..
        } => {
            fs::write(
                dir.join(COMPAT_FILE),
                format!(
                    "{SOURCE}\n// Over to the extension's own background code.\n\
                     importScripts({});\n",
                    json_string(path)
                ),
            )?;
        }
        Background::Scripts { .. } => {
            fs::write(dir.join(COMPAT_FILE), SOURCE)?;
        }
    }

    rewrite_background(&mut manifest, &background);
    manifest.insert(
        MANIFEST_KEY.to_string(),
        serde_json::json!({
            "added_files": added,
            "original_entry_point": entry,
        }),
    );

    // Written beside and renamed over, because a manifest cut off halfway by a
    // full disk is an extension nothing can read, and `fs::write` truncates
    // before it writes. The rename is within one directory, so it is atomic.
    let pending = dir.join(".manifest.json.incoming");
    let rewritten = serde_json::to_string_pretty(&manifest)
        .map_err(|e| ExtError::BadManifest(e.to_string()))?;
    fs::write(&pending, rewritten)?;
    fs::rename(&pending, &manifest_path)?;
    Ok(())
}

/// What the extension's `background` key declares, or `None` for a shape this
/// cannot get in front of.
fn read_background(value: Option<&serde_json::Value>) -> Option<Background> {
    let background = value?.as_object()?;

    if let Some(path) = background.get("service_worker").and_then(|v| v.as_str()) {
        if !is_package_relative(path) {
            return None;
        }
        return Some(Background::Worker {
            path: path.to_string(),
            module: background.get("type").and_then(|v| v.as_str()) == Some("module"),
        });
    }

    // MV2, and MV3's `scripts` fallback for browsers without workers. Only the
    // first entry is named in the notice — it is where the extension's code
    // begins, which is the fact a person needs — and the whole list keeps
    // running, in order, after ours.
    let scripts = background.get("scripts")?.as_array()?;
    let first = scripts.first()?.as_str()?;
    if !is_package_relative(first) {
        return None;
    }
    Some(Background::Scripts {
        first: first.to_string(),
    })
}

/// Whether a path names a file inside this package and nothing else.
///
/// The manifest is hostile input (ADR-0024) and this string is about to become
/// a JavaScript module specifier. Absolute paths, parent traversal and anything
/// carrying a scheme are refused, and refusing means the package is installed
/// untouched rather than installed with a specifier nobody vetted.
fn is_package_relative(path: &str) -> bool {
    !path.is_empty()
        && !path.starts_with('/')
        && !path.starts_with('\\')
        && !path.contains(':')
        && !path.contains('\\')
        && !path.split('/').any(|segment| segment == "..")
}

/// Point `background` at our file, keeping everything else the manifest said.
fn rewrite_background(manifest: &mut serde_json::Map<String, serde_json::Value>, of: &Background) {
    let Some(serde_json::Value::Object(background)) = manifest.get_mut("background") else {
        return;
    };
    match of {
        Background::Worker { .. } => {
            background.insert(
                "service_worker".to_string(),
                serde_json::Value::String(COMPAT_FILE.to_string()),
            );
        }
        Background::Scripts { .. } => {
            let Some(serde_json::Value::Array(scripts)) = background.get_mut("scripts") else {
                return;
            };
            scripts.insert(0, serde_json::Value::String(COMPAT_FILE.to_string()));
        }
    }
}

/// A path from the manifest as a JavaScript string literal.
///
/// Through `serde_json` rather than by wrapping it in quotes, so a manifest
/// that puts a quote or a newline in its own entry point produces a string
/// rather than a syntax error or a line of somebody else's JavaScript running
/// ahead of every extension this browser installs.
fn json_string(path: &str) -> String {
    serde_json::Value::String(path.to_string()).to_string()
}

/// The same, as a module specifier: relative, so it resolves against the
/// package root rather than against an import map nobody wrote.
fn module_specifier(path: &str) -> String {
    let literal = json_string(path);
    format!("\"./{}", &literal[1..])
}

#[cfg(test)]
#[path = "compat_tests.rs"]
mod tests;
